pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go

```diff
@@ -274,50 +277,72 @@ func (n *klusterletController) sync(ctx context.Context, controllerContext facto
 		managedClusterClients, err = n.buildManagedClusterClientsHostedMode(ctx, n.kubeClient, config.AgentNamespace, config.ExternalManagedKubeConfigSecret)
 		if err != nil {
-			_, _, _ = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(metav1.Condition{
+			newStatus, _, updateErr = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(metav1.Condition{
 				Type: klusterletReadyToApply, Status: metav1.ConditionFalse, Reason: "KlusterletPrepareFailed",
 				Message: fmt.Sprintf("Failed to build managed cluster clients: %v", err),
 			}))
-			return err
 		} else {
-			_, _, _ = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(metav1.Condition{
+			newStatus, _, updateErr = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(metav1.Condition{
 				Type: klusterletReadyToApply, Status: metav1.ConditionTrue, Reason: "KlusterletPrepared",
 				Message: "Klusterlet is ready to apply",
 			}))
 		}
+		if updateErr != nil {
+			return updateErr
+		}
+		klusterlet.Status = *newStatus
 	}
 
-	// Update finalizer after the clients created, otherwise, for Hosted mode if users did not provide the
-	// external managed cluster kubeconfig, we can not delete the finalizer
-	if klusterlet.DeletionTimestamp.IsZero() {
-		hasFinalizer := false
-		for i := range klusterlet.Finalizers {
-			if klusterlet.Finalizers[i] == klusterletFinalizer {
-				hasFinalizer = true
-				break
+	// Klusterlet is deleting, we remove its related resources on managed and management cluster
+	if !klusterlet.DeletionTimestamp.IsZero() {
+		var removeManagedClusterResources = true
+
+		if config.InstallMode == operatorapiv1.InstallModeHosted {
+			if hasFinalezer(klusterlet, klusterletHostedFinalizer) {
+				if !meta.IsStatusConditionTrue(klusterlet.Status.Conditions, klusterletReadyToApply) {
+					// wait for the external managed kubeconfig to exist to clean up resources on the manged cluster
+					return nil
+				}
+			} else {
+				removeManagedClusterResources = false
 			}
 		}
-		if !hasFinalizer {
-			klusterlet.Finalizers = append(klusterlet.Finalizers, klusterletFinalizer)
-			_, err = n.klusterletClient.Update(ctx, klusterlet, metav1.UpdateOptions{})
+
+		if err := n.cleanUp(ctx, controllerContext,
+			managedClusterClients, config, removeManagedClusterResources); err != nil {
 			return err
 		}
+
+		return n.removeKlusterletFinalizers(ctx, klusterlet)
 	}
 
-	// Klusterlet is deleting, we remove its related resources on managed cluster
-	if !klusterlet.DeletionTimestamp.IsZero() {
-		if err := n.cleanUp(ctx, controllerContext, managedClusterClients, config); err != nil {
-			return err
+	if config.InstallMode == operatorapiv1.InstallModeHosted {
+		if !meta.IsStatusConditionTrue(klusterlet.Status.Conditions, klusterletReadyToApply) {
```

Comment:
why this cannot be put in 291?

Reply:
if you put there, you do not need to check this condition. Only need to check the validity of the client, right?